File: src/algorithms/arrays/remove-duplicates/sources/RemoveDuplicates.java
// Remove Duplicates from Sorted Array — O(n) two-pointer: write pointer tracks unique boundary
public class RemoveDuplicates {
    public static int[] removeDuplicates(int[] sortedArray) {
        if (sortedArray.length == 0) { // @step:initialize
            return new int[]{0}; // @step:initialize
        }

        int[] result = sortedArray.clone();
        int writePointer = 0; // @step:initialize

        for (int readPointer = 1; readPointer < result.length; readPointer++) {
            if (result[readPointer] != result[writePointer]) { // @step:compare
                writePointer++; // @step:swap
                result[writePointer] = result[readPointer]; // @step:swap
            }
        }

        int uniqueCount = writePointer + 1;
        int[] uniqueResult = new int[uniqueCount + 1];
        uniqueResult[0] = uniqueCount;
        for (int copyIndex = 0; copyIndex < uniqueCount; copyIndex++) {
            uniqueResult[copyIndex + 1] = result[copyIndex]; // @step:complete
        }
        return uniqueResult; // @step:complete
    }
}

File: src/algorithms/arrays/two-pointer/remove-duplicates/sources/remove-duplicates.go
// Remove Duplicates from Sorted Array — O(n) two-pointer: write pointer tracks unique boundary
package removeduplicates

func removeDuplicates(sortedArray []int) (int, []int) {
	if len(sortedArray) == 0 {
		// @step:initialize
		return 0, []int{} // @step:initialize
	}

	result := make([]int, len(sortedArray))
	copy(result, sortedArray)
	writePointer := 0 // @step:initialize

	for readPointer := 1; readPointer < len(result); readPointer++ {
		if result[readPointer] != result[writePointer] { // @step:compare
			writePointer++                             // @step:swap
			result[writePointer] = result[readPointer] // @step:swap
		}
	}

	uniqueCount := writePointer + 1
	return uniqueCount, result[:uniqueCount] // @step:complete
}
